Simplify setup guide (#40)
sse-secure-systems/connaisseur: setup/README.md

1. **Local setup ([Minikube](https://github.com/kubernetes/minikube) & [Harbor](https://github.com/goharbor/harbor))**: Connaisseur can be run completely locally with a minikube instance paired with a local Harbor installation that combines registry and notary servers. This setup is a bit more cumbersome since it does not use existing public infrastructure such as PKI. However, it is entirely based on open-source solutions, does not require continuous access to the internet and you control every part of the installation. This makes it specifically suited e.g. for security researchers and users interested in the inner workings of DCT and Connaisseur. Head over to [the respective README](local/README.md) to try it out.

2. **[Azure Kubernetes Services](https://docs.microsoft.com/en-us/azure/aks/) with [Azure Container Registry (ACR)](https://docs.microsoft.com/en-us/azure/container-registry/)**: Microsoft Azure is currently the only one of the big cloud providers with a managed Kubernetes and container registry that supports DCT. If you want to set up Connaisseur with ACR, head over to [the specialized setup guide](acr/README.md).

## Requirements

> We assume you have a running Kubernetes cluster already. Since this tutorial works with the CLI, we furthermore assume you have the `docker`, `git`, `helm`, `kubectl`, `make`, `openssl` and `yq` binaries installed and usable, i.e. having run `docker login` and having switched to the appropriate `kubectl` context. 

Reviewer: ```suggestion
We assume you have a running Kubernetes cluster already. Since this tutorial works with the CLI, we furthermore assume you have the `docker`, `git`, `helm`, `kubectl`, `make`, `openssl` and `yq` binaries installed and usable, i.e. having run `docker login` and having switched to the appropriate `kubectl` context.
```

- Generate the corresponding public key via `openssl ec -in root-priv.key -pubout -out root.pub`. You will be asked to enter your root password.
- The new `root.pub` contains your public which you copy and set for `notary.rootPubKey` in `connaisseur/helm/values.yaml`. The result should look similar as for the first case above.
- To clean up, remove the `root-priv.key` and `root.pub` in `~/.docker/trust/private`.

##### bash

- Run the script below from the Connaisseur repository to generate your root public key. You will be asked to enter your root password set above:

```bash
FIRST_LINE=$(grep -n rootPubKey: helm/values.yaml | sed s/:.*//)
LAST_LINE=$(grep -nF -- '-----END PUBLIC KEY-----' helm/values.yaml | sed s/:.*//)
sed -i $((${FIRST_LINE} + 1)),${LAST_LINE}d helm/values.yaml

cd ~/.docker/trust/private
sed '/^role:\sroot$/d' $(grep -iRl "role: root$" .) > root-priv.key
openssl ec -in root-priv.key -pubout -out root.pub
```

- After entering your password, copy the public key to the `helm/values.yaml`: 

```bash
sed -i "${FIRST_LINE}s?.*?  rootPubKey: |\n    $(sed ':a;N;$!ba;s/\n/\\n    /g' root.pub)?" ${OLDPWD}/helm/values.yaml
rm root-priv.key root.pub
cd -
```

> - The resulting `/helm/values.yaml` should look similar as for the first case above.

Reviewer: ```suggestion
- The resulting `/helm/values.yaml` should have an excerpt looking similar to what is shown in the first case above.
```

`kubectl get all -n connaisseur` will show you most* of what you deployed (ConfigMaps, Secrets and MutatingWebhookConfiguration are not shown).

### Deploy (un)signed images

Let's test if our configuration works. We need to prepare a signed and an unsigned image and push these to a remote repository. For simplicity, we define an image path variable (host, repository and image). Below, substitute `IMAGE_PATH` with your own the values as appropriate:

```bash
# Typically, IMAGE_PATH=<YOUR-REGISTRY>/<YOUR-REPOSITORY-NAME-/-DOCKER-HUB-ID>/<IMAGE-NAME>

> IMAGE_PATH=docker.io/connytest/testimage

Reviewer 2: Who is the owner of this repo?

Author: Me. It is also the one I use in the demo. It all ties together and you are safe if you use that value and the sample key below ;-)
Also created #41 about this. What is bad if we move it into SSE, is that it makes every command like this much longer...just from a UX perspective.
What do you think?

Reviewer 2: Oh, didn't see that :+1: I think it would be cleaner if it was moved into SSE. Also, I don't think the long path matters much. However, if we decide to move it we can do it independently of this PR I guess :)

Author: yes, that was my rationale as well...

  Repository Key:   130b5abbea417fea7e2a0acd2cc0a3a84f81d5b763ed82dcfaad8dceebac0b75
  Root Key:   6b35860633a0cf852670fd9b5c12ba068875f3804d6711feb16fcd74c723c816
```

Note that there is only trust data for the `signed` tag, not for `unsigned`. Let's try to deploy our fresh images to the cluster. We start with the signed image:

```bash
kubectl run signed --image=${IMAGE_PATH}:signed
```

You should be prompted that the deployment was successful:

```bash
pod/signed created
```

Finally, let's make sure that Connaisseur will not just allow deployment of any image:

```bash
kubectl run unsigned --image=${IMAGE_PATH}:unsigned
```

You should see the deployment being rejected with an error similar to:

```bash
Error from server: admission webhook "connaisseur-svc.connaisseur.svc" denied the request: could not find signed digest for image "docker.io/connytest/testimage:unsigned" in trust data.
```

or if you pushed the unsigned image to another registry altogether:

```bash
Error from server: admission webhook "connaisseur-svc.connaisseur.svc" denied the request: no trust data for image "docker.io/connytest/testimage:unsigned".
```

> > Note that while deployment of containers is blocked Kubernetes will still create services or other resources you might specify in a deployment.yaml, because those does not reference an image and are thus not denied by Connaisseur. You might have to clean up denied deployments.

Reviewer: ```suggestion
> Note that while deployment of containers is blocked Kubernetes will still create services or other resources you might specify in a deployment.yaml, because those do not reference an image and are thus not denied by Connaisseur. You might have to clean up denied deployments.
```

    port: 8080
    targetPort: 5000
  selector:
    app: sample
  type: LoadBalancer
EOF
kubectl run unsigned --image=${IMAGE_PATH}:unsigned
```

This deployment should be accepted with
You should see the deployment being rejected with an error similar to:

```bash
deployment.apps/sample-deployment created
service/sample created
Error from server: admission webhook "connaisseur-svc.connaisseur.svc" denied the request: could not find signed digest for image "docker.io/connytest/testimage:unsigned" in trust data.
```

Your signed images were allowed through, in contrast to those unsigned ones with unknown content. DCT, not in a nutshell, but in Kubernetes!
or if you pushed the unsigned image to another registry altogether:

```bash
Error from server: admission webhook "connaisseur-svc.connaisseur.svc" denied the request: no trust data for image "docker.io/connytest/testimage:unsigned".
```

### 2. Cleanup
> Note that while deployment of containers is blocked Kubernetes will still create services or other resources you might specify in a deployment.yaml, because those does not reference an image and are thus not denied by Connaisseur. You might have to clean up denied deployments.

> If you want to remove the test deployment of step 1, run `kubectl delete all -lapp=sample`.

Reviewer: Would still keep this deletion of test resources, maybe in the cleanup section

Author: but there is no real resources anymore, just two pods. Will add that below though...

Author: done!
